packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts

```diff
@@ -585,6 +623,146 @@ export const samplerOperations = {
+                                    this.getUniswapV2SellQuotes(
+                                        [takerToken, wethAddress, makerToken],
+                                        takerFillAmounts,
+                                    ),
+                                );
+                            }
+                            return ops;
+                        case ERC20BridgeSource.Kyber:
+                            return this.getKyberSellQuotes(makerToken, takerToken, takerFillAmounts);
+                        case ERC20BridgeSource.Curve:
+                            return getCurveInfosForPair(takerToken, makerToken).map(curve =>
+                                this.getCurveSellQuotes(
+                                    curve,
+                                    curve.tokens.indexOf(takerToken),
+                                    curve.tokens.indexOf(makerToken),
+                                    takerFillAmounts,
+                                ),
+                            );
+                        case ERC20BridgeSource.LiquidityProvider:
+                            if (liquidityProviderRegistryAddress === undefined) {
+                                throw new Error(
+                                    'Cannot sample liquidity from a LiquidityProvider liquidity pool, if a registry is not provided.',
+                                );
+                            }
+                            return this.getLiquidityProviderSellQuotes(
+                                liquidityProviderRegistryAddress,
+                                makerToken,
+                                takerToken,
+                                takerFillAmounts,
+                            );
+                        case ERC20BridgeSource.MultiBridge:
+                            if (multiBridgeAddress === undefined) {
+                                throw new Error(
+                                    'Cannot sample liquidity from MultiBridge if an address is not provided.',
+                                );
+                            }
+                            const intermediateToken = getMultiBridgeIntermediateToken(takerToken, makerToken);
+                            return this.getMultiBridgeSellQuotes(
+                                multiBridgeAddress,
+                                makerToken,
+                                intermediateToken,
+                                takerToken,
+                                takerFillAmounts,
+                            );
+                        case ERC20BridgeSource.Balancer:
+                            return this.balancerPoolsCache
+                                .getCachedPoolAddressesForPair(takerToken, makerToken)
+                                .map(poolAddress =>
+                                    this.getBalancerSellQuotes(poolAddress, makerToken, takerToken, takerFillAmounts),
+                                );
```

Comment:
note that this will _only_ be used for two-hop quotes.
Recall that `ERC20BridgeSource.Balancer` gets excluded from the sources passed into `getSellQuotes` and instead we use the BalancerPoolCache/GraphQL query to compute Balancer single-hop quotes. 
For two-hop, the balancer pool addresses need to be known _before_ the sampler RPC call. I want to avoid performing the GraphQL query + RPC call synchronously, so instead we just pull whatever pool addresses are in the cache and pass those to the sampler.

Reply:
We could also adopt this approach of using BalancerPoolsCache/GraphQL to fetch pool addresses and BalancerSampler to fetch balances/compute quotes for single-hop quotes, but then we would also potentially need to make a GraphQL query + RPC call synchronously (though we'd dial the cache expiry way up, so the GraphQL queries would be less frequent)